[13.0][MIG+IMP] account_invoice_date_due: Migration to 13.0 and refactor

joao-p-marques opened the pull request:
1. Migration
2. Refactor to use an editable field instead of a wizard
3. Add tests

@Tecnativa
TT28875

ping @victoralmau @pedrobaeza

luistorresm committed:
[ADD] account_invoice_date_due: Date Due now could be updated

Allow update the date due in an invoice (Even if it's done). To avoid
cancel the invoice and generate again

luisg123v committed:
[FIX] account_invoice_date_due: Small fixes to the initial version

This applies pending comments on #632, which are:
- Fix some typos
- Be consistent with the license, use AGPL-3
- Add maintainer (not commented)
- Set group data as noupdate (not commented)

marylla committed:
Added translation using Weblate (German)

marylla committed:
Translated using Weblate (German)

Currently translated at 100.0% (15 of 15 strings)

Translation: account-invoicing-12.0/account-invoicing-12.0-account_invoice_date_due
Translate-URL: https://translation.odoo-community.org/projects/account-invoicing-12-0/account-invoicing-12-0-account_invoice_date_due/de/

anasuarez1 committed:
Translated using Weblate (Spanish)

Currently translated at 100.0% (15 of 15 strings)

Translation: account-invoicing-12.0/account-invoicing-12.0-account_invoice_date_due
Translate-URL: https://translation.odoo-community.org/projects/account-invoicing-12-0/account-invoicing-12-0-account_invoice_date_due/es/

joao-p-marques committed:
[IMP] account_invoice_date_due: black, isort, prettier

joao-p-marques committed:
[MIG] account_invoice_date_due: Migration to 13.0

TT28875

victoralmau commented on account_invoice_date_due/models/__init__.py:
```diff
@@ -0,0 +1,3 @@
+# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl).
+
+from . import update_due_date
```
```suggestion
from . import account_move
```

chienandalu reviewed (commented):
There's a travis error. I'm not sure it's related to this PR though...

joao-p-marques commented:
> There's a travis error. I'm not sure it's related to this PR though...

Thanks @chienandalu Indeed, it was due to the `onchange` I added, should be fixed now.

joao-p-marques commented:
@chienandalu I think I have fixed the handling of the `invisible`/`readonly` situations, integrating the standard Odoo mechanism for handling those fields. Can you check, please?

chienandalu commented on account_invoice_date_due/views/account_move_date_due.xml:
```diff
@@ -0,0 +1,20 @@
+                <field name="is_invoice_date_due_editable" invisible="1" />
+            </field>
+            <field name="invoice_date_due" position="attributes">
+                <attribute name="readonly">0</attribute>
```
Wouldn't it be easier to play with fields declared multiple times and positive/negative groups? Something like this for the regular field:
```xml
<attribute name="groups">!account_invoice_date_due.allow_to_change_due_date</attribute>
```
And then:
```xml
<field name="invoice_date_due" position="after">
    <field name="invoice_date_due" groups="account_invoice_date_due.allow_to_change_due_date" attrs="{'readonly': your domain, 'invisible': another domain}" />
</field>
```

chienandalu commented on account_invoice_date_due/models/account_move.py:
```diff
@@ -0,0 +1,28 @@
+            in self.env.ref("account_invoice_date_due.allow_to_change_due_date").users
+        )
+        for rec in self:
+            if user_in_group and rec.state == "draft":
```
This is wrong, isn't it?

joao-p-marques commented:
Thank you for the suggestions @chienandalu 
I think now it is correct :+1: 

I just have a question: should we make the `invoice_payment_term_id` field also editable after the invoice has been posted? The date field will be invisible if a payment term is set, so it wouldn't be possible to change the value.

chienandalu commented:
> I just have a question: should we make the invoice_payment_term_id field also editable after the invoice has been posted? The date field will be invisible if a payment term is set, so it wouldn't be possible to change the value.

What could you do in the previous version?

joao-p-marques commented:
> What could you do in the previous version?

With the wizard, you could change the due date even if a different payment term had been set.

pedrobaeza commented:
It's a difficult question, as letting the payment term to be editable will make very difficult to match previous AR/AP lines with the corresponding due dates, but also not letting to edit anything will make this to have less features.

Can you make due date always editable once confirmed no matter the initial data and propagate the date?

joao-p-marques commented:
Yes, but the problem is keeping the Odoo mechanism to hide the Due date 
field once a payment term is selected... You need to remove that 
payment term to further edit the due date field.

On qui, mar 25, 2021 at 03:04, Pedro M. Baeza 
***@***.***> wrote:
> It's a difficult question, as letting the payment term to be editable 
> will make very difficult to match previous AR/AP lines with the 
> corresponding due dates, but also not letting to edit anything will 
> make this to have less features.
> 
> Can you make due date always editable once confirmed no matter the 
> initial data and propagate the date?
> 
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub 
> <https://github.com/OCA/account-invoicing/pull/895#issuecomment-806520626>, 
> or unsubscribe 
> <https://github.com/notifications/unsubscribe-auth/AJJMDDRRQNCCLMDW3PHZW6TTFMDDTANCNFSM4ZS6C37Q>.
>

chienandalu commented:
> Yes, but the problem is keeping the Odoo mechanism to hide the Due date field once a payment term is selected... You need to remove that payment term to further edit the due date field.


But you're using a diferent field declaration in the view for the proper group, so it should not affect it, or does it?

joao-p-marques commented:
> But you're using a diferent field declaration in the view for the proper group, so it should not affect it, or does it?

Yes, but I kept the declaration of the `invisible` attribute on both of them, to keep the original mechanism. Another option is to make it always visible in another field in the view.

joao-p-marques commented:
@chienandalu @pedrobaeza I just added some changes with a slightly different flow, which IMHO is more clear. It is documented under USAGE and added in the tests. Can you check again please?

chienandalu commented on account_invoice_date_due/models/account_move.py:
```diff
@@ -0,0 +1,24 @@
+    def _onchange_invoice_date_due(self):
+        # Set due date on account move lines
+        if self.invoice_date_due:
+            self.mapped("line_ids").date_maturity = self.invoice_date_due
```
In my functional test, I changed the due date of the invoice but the maturity date wasn't set. Any way, shouldn't you set it only for the debit lines?

joao-p-marques commented:
@chienandalu changes done

chienandalu commented on account_invoice_date_due/models/account_move.py:
```diff
@@ -0,0 +1,30 @@
+        res = super().write(vals)
+        # Propagate due date to move lines
+        # that correspont to the receivable/payable account
+        if "invoice_date_due" in vals and self.state in ("draft", "posted"):
```
I'd say you won't have payment term lines when the invoice is in draft
```suggestion
        if "invoice_date_due" in vals and self.state == "posted":
```

joao-p-marques replied:
Yes, it was just that, from my testing, the correspondent move lines that should be updated are created already while in draft mode. 
Should I just ignore the values while in draft and change only when posted?

chienandalu replied:
The thing is that when you're in draft you'll work with the standard behavior. The edition of the due date should be allowed after posting the invoice

joao-p-marques replied:
Ok, I think I was looking at the problem wrong :sweat_smile: 
I removed that condition following your suggestion and also removed the check from the tests (I now only check the propagation after having posted the invoice)

chienandalu reviewed (commented):
Just a little detail

joao-p-marques committed:
[IMP] account_invoice_date_due: Refactor and add tests

Don't use the wizard, but make the invoice_date_due editable for a specific group
Still, propagate the change to the invoice lines
Add tests

TT28875

pedrobaeza reviewed (changes requested):
Trying on runbot, I have posted an invoice, and I'm not able to edit the due date:

![Selección_007](https://user-images.githubusercontent.com/7165771/113388458-9044d000-938e-11eb-8b5a-c870a1f709f8.png)

joao-p-marques commented:
@pedrobaeza You need to add the user to the group that allows editing that field, as happened in v12:
![image](https://user-images.githubusercontent.com/38977934/113540949-1a807480-95d9-11eb-8a9c-bd7704885b6c.png)

I've tested again in the runbot and everything seems to work

pedrobaeza reviewed (approved):
Oh, stupid mistake not seeing the group.

pedrobaeza commented:
/ocabot merge nobump
